File: y2024/day11/day11.go
package day11

import (
	"adventofcode/y2024/utils"
	"fmt"
	"math"
	"strconv"
	"strings"
)

func processInput(daydir string) map[int]int {
	content := utils.Read(daydir + "/input.txt")
	content = strings.TrimPrefix(content, "\n")
	splitContent := strings.Split(content, " ")
	retval := map[int]int{}
	for _, numbah := range splitContent {
		n, _ := strconv.Atoi(numbah)
		if _, ok := retval[n]; ok {
			retval[n] += 1
		} else {
			retval[n] = 1
		}
	}
	if _, ok := retval[0]; !ok {
		retval[0] = 0
	}
	if _, ok := retval[1]; !ok {
		retval[1] = 0
	}
	return retval
}

func getNumDigits(num int) int {
	return int(math.Floor(math.Log10(float64(num)) + 1))
}

func CountRocks(rocks *map[int]int) int {
	total := 0

	for _, v := range *rocks {
		total += v
	}

	return total
}

func part01(rocks map[int]int, numBlinks int, part string) {
	for range numBlinks {
		newRocks := map[int]int{}
		// fmt.Printf(" len(newRocks): %d\n", len(newRocks))
		// fmt.Printf("%v, ", rocks)
		for r, count := range rocks {
			numDigits := getNumDigits(r)
			if r == 0 {
				if _, ok := newRocks[1]; !ok {
					newRocks[1] = 0
				}
				newRocks[1] += count
			} else if numDigits%2 == 0 {
				tenPow := numDigits / 2
				left := r / int(math.Pow10(tenPow))
				right := r % int(math.Pow10(tenPow))
				if _, ok := newRocks[left]; !ok {
					newRocks[left] = 0
				}
				if _, ok := newRocks[right]; !ok {
					newRocks[right] = 0
				}
				newRocks[left] += count
				newRocks[right] += count
			} else {
				if _, ok := newRocks[r]; !ok {
					newRocks[r] = 0
				}
				newRocks[r*2024] += count
			}
		}
		rocks = newRocks
	}
	fmt.Printf("%s: %d\n", part, CountRocks(&rocks))
}

func Run(dir string) {
	utils.PartPrinter("DAY 11")
	data := processInput(dir + "/day11")
	part01(data, 25, "PART 01")
	part01(data, 75, "PART 02")
}

File: y2024/day11/day11.py
import math
from copy import deepcopy


INPUT = "17639 47 3858 0 470624 9467423 5 188"


def process_input():
    with open("./day11/input.txt", "r") as fp:
        content = fp.read()
    content = content.strip()
    content = {int(i):1 for i in content.split()}
    if content.get(1) is None:
        content[1] = 0
    if content.get(0) is None:
        content[0] = 0
    return content


def get_num_digits(n: int) -> int:
    return math.floor(math.log10(n)+ 1)


def part01(data: dict[int, int], num_blinks: int):
    rocks = deepcopy(data)
    for i in range(num_blinks):
        new_rocks = {}
        for stone, count in rocks.items():
            if stone == 0:
                if new_rocks.get(1) is None:
                    new_rocks[1] = 0
                new_rocks[1] += count
            elif get_num_digits(stone)%2==0:
                num_digits = get_num_digits(stone)
                ten_pow = num_digits // 2
                left = stone // 10**ten_pow
                right = stone % 10**ten_pow
                if new_rocks.get(left) is None:
                    new_rocks[left] = 0
                if new_rocks.get(right) is None:
                    new_rocks[right] = 0
                new_rocks[left] += count
                new_rocks[right] += count
            else:
                if new_rocks.get(stone*2024) is None:
                    new_rocks[stone*2024] = 0
                new_rocks[stone * 2024] += count
        rocks = new_rocks
        print(f"blinking: {i} - num_rocks = {sum(rocks.values())}")
    print(f"num rocks: {sum(rocks.values())}")


def main():
    data = process_input()
    part01(data, 25)
    part01(data, 75)


if __name__ == "__main__":
    main()

